File: cpp/111. 二叉树的最小深度.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int minDepth(TreeNode* root) {
        if(root == nullptr) return 0;
        queue<TreeNode*> qu;
        qu.push(root);
        int depth = 1;
        while(!qu.empty()){
            int sz = qu.size();
            for(int i=0; i<sz; i++) {
                TreeNode* node = qu.front();
                qu.pop();
                if(node->left == nullptr && node->right == nullptr) {
                    return depth;
                }
                if(node->left != nullptr) {
                    qu.push(node->left);
                }
                if(node->right !=nullptr) {
                    qu.push(node->right);
                }
            }
            depth++;
        }
        return depth;
    }
};

File: 刷题/golang/111. 二叉树的最小深度.go
func minDepth(root *TreeNode) int {
	if root == nil {
		  return 0
	 }
	 q := []*TreeNode{root}
	 // root 本身就是一层，depth 初始化为 1
	 depth := 1
 
	 for len(q) != 0 {
		 sz := len(q)
		 /* 将当前队列中的所有节点向四周扩散 */
		 for i := 0; i < sz; i++ {
			 cur := q[0]
			 q = q[1:]
			 /* 判断是否到达终点 */
			 if cur.Left == nil && cur.Right == nil {
				 return depth
			 }
			 /* 将 cur 的相邻节点加入队列 */
			 if cur.Left != nil {
				 q = append(q, cur.Left)
			 }
			 if cur.Right != nil {
				 q = append(q, cur.Right)
			 }
		 }
		 /* 这里增加步数 */
		 depth++
	 }
	 return depth
 }
 